subsys/usb/device_next/class/bt_hci.c

```diff
@@ -0,0 +1,491 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Bluetooth HCI USB transport layer implementation with following
+ * endpoint configuration:
+ *  - HCI commands through control endpoint (host-to-device only)
+ *  - HCI events through interrupt IN endpoint
+ *  - ACL data through one bulk IN and one bulk OUT endpoints
+ *
+ * TODO: revise transfer handling so that Bluetooth net_bufs can be used
+ * directly without intermediate copying.
+ *
+ * Limitations:
+ *  - Remote wakeup before IN transfer is not yet supported.
+ *  - H4 transport layer is not yet supported
+ */
+
+#include <zephyr/init.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/usb/usbd.h>
+#include <zephyr/usb/usb_ch9.h>
+#include <zephyr/drivers/usb/udc.h>
+
+#include <zephyr/net/buf.h>
+
+#include <zephyr/bluetooth/buf.h>
+#include <zephyr/bluetooth/hci_raw.h>
+#include <zephyr/bluetooth/l2cap.h>
+#include <zephyr/bluetooth/hci_vs.h>
+#include <zephyr/drivers/bluetooth/hci_driver.h>
+#include <zephyr/sys/atomic.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_hci, CONFIG_USBD_BT_HCI_LOG_LEVEL);
+
+#define USB_BT_HCI_SUBCLASS		0x01
+#define USB_BT_HCI_PROTOCOL		0x01
+
+#define BT_HCI_DEFAULT_BULK_EP_MPS	0	/* Get maximum supported */
+#define BT_HCI_DEFAULT_INT_EP_MPS	64
+#define BT_HCI_CLASS_ENABLED		0
+#define BT_HCI_ACL_RX_ENGAGED		1
+
+static K_FIFO_DEFINE(bt_hci_rx_queue);
+static K_FIFO_DEFINE(bt_hci_tx_queue);
+
+/*
+ * REVISE: global (bulk, interrupt, iso) specific pools would be more
+ * RAM usage efficient.
+ */
+NET_BUF_POOL_FIXED_DEFINE(bt_hci_ep_pool,
+			  3, 512,
+			  sizeof(struct udc_buf_info), NULL);
+/* HCI RX/TX threads */
+static K_KERNEL_STACK_DEFINE(rx_thread_stack, CONFIG_BT_HCI_TX_STACK_SIZE);
+static struct k_thread rx_thread_data;
+static K_KERNEL_STACK_DEFINE(tx_thread_stack, 512);
+static struct k_thread tx_thread_data;
+
+struct bt_hci_data {
+	struct net_buf *acl_buf;
+	uint16_t acl_len;
+	struct k_sem sync_sem;
+	atomic_t state;
+};
+
+/*
+ * Make supported device request visible for the stack
+ * bRequest 0x00 and bRequest 0xE0.
+ */
+static const struct usbd_cctx_vendor_req bt_hci_vregs =
+	USBD_VENDOR_REQ(0x00, 0xe0);
+
+struct usbd_bt_hci_desc {
+	struct usb_if_descriptor if0;
+	struct usb_ep_descriptor if0_int_ep;
+	struct usb_ep_descriptor if0_out_ep;
+	struct usb_ep_descriptor if0_in_ep;
```

Comment:
The order is wrong. See BLUETOOTH CORE SPECIFICATION Version 5.3 | Vol 4, Part B page 1667 Table 2.1 outlines a recommended configuration for a USB Full Speed device

The order should be:
HCI events Interrupt IN
ACL Data Bulk IN
ACL Data Bulk OUT

Moreover, there should also be two dummy isochronous endpoints (first Isoch IN, then Isoch OUT) with Max Packet Size 0 (i.e. no data can be transferred to/from these)

Reply:
Order should not matter, as you mention it "outlines a **recommended** configuration", the host have to parse device descriptors and interpret it correctly. I also see no benefit to add isochronous endpoints, there are none in subsys/usb/device/class/bluetooth.c and we do not support classic voice channels.

Reply:
As it turns out, Bluetooth specification versions prior to 4.0 had "required configuration" and starting with version 4.0 it is "recommended configuration". For earlier versions the Core Spec Addendum 2 changes the "required" to "recommended".